python/lsst/ci/tsort.py

```diff
@@ -1,59 +1,103 @@
+    """Takes an iterable collection of (object, dependency) pairs
+    and returns an list of ordered dependencies.
+    """
+    return toposort_dfs_mapping(to_dep_graph(edges))
+
+
+def toposort_mapping(graph_set: Mapping[str, Set[str]]) -> Iterator[List]:
+    """Returns an iterator of ordered dependency lists from a 
+    graph dictionary.
+    The items in each list can be processed in any order.
     """
-    # resulting list
-    L = []  # noqa
-
-    # maintain forward and backward edge maps in parallel.
-    st, ts = {}, {}
-
-    def prune(s, t):
-        del st[s][t]
-        del ts[t][s]
-
-    def add(s, t):
-        try:
-            st.setdefault(s, {})[t] = 1
-        except Exception as e:
-            raise RuntimeError(e, (s, t))
-        ts.setdefault(t, {})[s] = 1
-
-    for s, t in edges:
-        add(s, t)
-
-    # frontier
-    S = set(st.keys()).difference(ts.keys())  # noqa
-
-    while S:
-        s = S.pop()
-        L.append(s)
-        for t in list(st.get(s, {}).keys()):
-            prune(s, t)
-            if not ts[t]:       # new frontier
-                S.add(t)
-
-    if [_f for _f in st.values() if _f]:  # we have a cycle. report the cycle.
-        def traverse(vs, seen):
-            for s in vs:
-                if s in seen:
-                    raise GraphError('contains cycle: ', seen)
-                seen.append(s)  # xx use ordered set..
-                traverse(st[s].keys(), seen)
-        traverse(st.keys(), list())
-        assert False, 'should not reach..'
-
-    return L
+    all_dependencies = set()
+    self_including_nodes = []
+    for node, dependencies in graph_set.items():
+        if node in dependencies:
+            self_including_nodes.append(node)
+        all_dependencies.union(dependencies)
+    if self_including_nodes:
+        raise GraphError(f"Self-including nodes: {self_including_nodes}")
+    missing_nodes = list(sorted(all_dependencies - set(graph_set.keys())))
+    if missing_nodes:
+        raise GraphError(f"Missing nodes in mapping: {missing_nodes}")
+    remaining_nodes = graph_set.copy()
+    while True:
+        # Visit all nodes, remove nodes without dependencies
+        childless_nodes = set(node for node, dependencies in remaining_nodes.items() if not dependencies)
+        if not childless_nodes:
+            break
+        yield list(sorted(childless_nodes))
+        more_nodes = {}
+        # Remove nodes we returned from dependency lists
+        for node, dependencies in remaining_nodes.items():
+            if node not in childless_nodes:
+                more_nodes[node] = (dependencies - childless_nodes)
+        remaining_nodes = more_nodes
+    if remaining_nodes:
+        raise GraphError(f"Cycle among nodes: {remaining_nodes}")
+
+
+def toposort_dfs_mapping(graph: Mapping[str, Set[str]]) -> List[str]:
+    """Topological sort - Depth-first search.
+    The results are already flattened
+    """
+    class Node:
```

Comment:
Please document the parameters. I'm guessing dependencies is a collection of package names, but it would be nice to know.